File: go/AssemblerWeb/endpoint/security_validator.go
package endpoint

import (
	"assembler/config"
	"strings"
)

// Maximum parameter length to prevent DoS attacks
const ParamMaxLength = 256

// Maximum log file size (500 KB)
const MaxLogFileSize = 500 * 1024

// Output size buffer (50 KB)
const OutputSizeBuffer = 50 * 1024

// IsValidContentSize validates content size against a maximum limit (bytes)
func IsValidContentSize(content *string, maxSize int) bool {
	if content == nil || *content == "" {
		return true
	}
	return len([]byte(*content)) <= maxSize
}

// IsValidOutputSizeWithBuffer validates output size against template total size plus buffer
func IsValidOutputSizeWithBuffer(htmlContent *string, templateTotalSize int) bool {
	if htmlContent == nil || *htmlContent == "" {
		return true
	}
	outputSize := len([]byte(*htmlContent))
	if templateTotalSize > 0 {
		maxAllowedSize := templateTotalSize + OutputSizeBuffer
		return outputSize <= maxAllowedSize
	}
	// If template size is unknown, reject
	return false
}

// IsValidLogContent validates log content format and size (basic, no regex)
func IsValidLogContent(logContent *string) (bool, string) {
	if logContent == nil || *logContent == "" {
		return false, "Log content is empty"
	}
	if !IsValidContentSize(logContent, MaxLogFileSize) {
		return false, "Log file exceeds maximum size limit (500 KB)"
	}
	lines := strings.Split(*logContent, "\n")
	validLines := 0
	totalLines := 0
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		totalLines++
		// Accept lines that look like log entries or stack traces (basic)
		if strings.HasPrefix(line, "[") || strings.HasPrefix(line, "at ") || strings.HasPrefix(line, "\tat ") {
			validLines++
		} else if len(line) > 10 {
			validLines++ // Accept long lines as possible log messages
		}
	}
	if totalLines > 0 && float64(validLines)/float64(totalLines) < 0.5 {
		return false, "Log content does not match expected format"
	}
	return true, ""
}

// GetTemplateTotalSize gets the template total size for an AppSite/AppView from scenarios
func GetTemplateTotalSize(appSite, appView string) int {
	scenarios, err := config.GetScenarios()
	if err != nil {
		return 0
	}
	for _, s := range scenarios {
		if strings.EqualFold(s.AppSite, appSite) && strings.EqualFold(s.AppView, appView) {
			return s.TotalSize
		}
	}
	return 0
}

// ValidEngineTypes is the allowlist of valid engine types
var ValidEngineTypes = map[string]bool{
	"Normal":     true,
	"PreProcess": true,
}

// GetValidAppSites gets the valid AppSites from ConfigUtil
func GetValidAppSites() (map[string]bool, error) {
	return config.GetAppSites()
}

// IsValidPathComponent validates if a path component is safe
func IsValidPathComponent(value *string) bool {
	if value == nil {
		return false
	}

	v := strings.TrimSpace(*value)
	if v == "" {
		return false
	}

	// Check parameter length to prevent DoS
	if len(v) > ParamMaxLength {
		return false
	}

	// Check for path traversal attempts
	if strings.Contains(v, "..") || strings.Contains(v, "/") || strings.Contains(v, "\\") {
		return false
	}

	// Check for suspicious characters
	invalidChars := []rune{'<', '>', ':', '"', '|', '?', '*', '\x00'}
	for _, char := range v {
		for _, invalid := range invalidChars {
			if char == invalid {
				return false
			}
		}
		// Check for control characters
		if char < 32 {
			return false
		}
	}

	return true
}

// IsValidEngineType validates engine type against allowlist (case-insensitive)
func IsValidEngineType(engineType string) bool {
	for validType := range ValidEngineTypes {
		if strings.EqualFold(validType, engineType) {
			return true
		}
	}
	return false
}

// IsValidAppSite validates app_site against allowlist (case-insensitive)
func IsValidAppSite(appSite string, validAppSites map[string]bool) bool {
	for validSite := range validAppSites {
		if strings.EqualFold(validSite, appSite) {
			return true
		}
	}
	return false
}

File: node/AssemblerWeb/endpoint/securityValidator.js
import fsSync from 'fs';

// Import ConfigUtil dynamically based on environment
const assemblerBasePath = fsSync.existsSync('/app/wwwroot') ? '../Assembler/src' : '../../Assembler/src';
const configModule = await import(`${assemblerBasePath}/config/index.js`);
const { ConfigUtil } = configModule;

// Maximum parameter length to prevent DoS attacks
const PARAM_MAX_LENGTH = 256;

// Valid engine types allowlist
const VALID_ENGINE_TYPES = new Set(['normal', 'preprocess']);

// Cached ValidAppSites - loaded on first request
let cachedValidAppSites = null;

/**
 * Gets the valid AppSites from ConfigUtil. Throws if not loaded.
 */
export function getValidAppSites() {
  if (cachedValidAppSites !== null) {
    return cachedValidAppSites;
  }

  cachedValidAppSites = ConfigUtil.getAppSites();
  return cachedValidAppSites;
}

/**
 * Clears the AppSites cache - useful for development/testing
 */
export function clearAppSitesCache() {
  cachedValidAppSites = null;
}

/**
 * Validates if a path component is safe (no traversal, invalid chars, or excessive length)
 */
export function isValidPathComponent(value) {
  if (!value || typeof value !== 'string') {
    return false;
  }

  const v = value.trim();
  if (v === '') {
    return false;
  }

  // Check parameter length to prevent DoS
  if (v.length > PARAM_MAX_LENGTH) {
    return false;
  }

  // Check for path traversal attempts
  if (v.includes('..') || v.includes('/') || v.includes('\\')) {
    return false;
  }

  // Check for other suspicious characters
  const invalidChars = ['<', '>', ':', '"', '|', '?', '*', '\0'];
  for (const char of v) {
    if (invalidChars.includes(char)) {
      return false;
    }
    // Check for control characters
    if (char.charCodeAt(0) < 32) {
      return false;
    }
  }

  return true;
}

/**
 * Validates engine type against allowlist (case-insensitive)
 */
export function isValidEngineType(engineType) {
  if (!engineType || typeof engineType !== 'string') {
    return false;
  }
  return VALID_ENGINE_TYPES.has(engineType.toLowerCase());
}

/**
 * Validates app_site against allowlist (case-insensitive)
 */
export function isValidAppSite(appSite, validAppSites) {
  if (!appSite || typeof appSite !== 'string') {
    return false;
  }
  return validAppSites.has(appSite.toLowerCase());
}

// Maximum content sizes to prevent DDOS attacks
export const MAX_LOG_FILE_SIZE = 500 * 1024; // 500 KB
// Buffer allowance for output size validation (50 KB)
export const OUTPUT_SIZE_BUFFER = 50 * 1024;

/**
 * Validates log content format and size similar to C# implementation
 * @param {string} logContent
 * @returns {{valid:boolean, errorMessage?:string}}
 */
export function isValidLogContent(logContent) {
  if (!logContent || typeof logContent !== 'string') {
    return { valid: false, errorMessage: 'Log content is empty' };
  }

  if (Buffer.byteLength(logContent, 'utf8') > MAX_LOG_FILE_SIZE) {
    return { valid: false, errorMessage: 'Log file exceeds maximum size limit (500 KB)' };
  }

  const pattern = /^[\[\]0-9:\-\s\.TZ]+\s*(DEBUG|INFO|WARN|ERROR|TRACE|FATAL)?:?\s*.+$/im;
  const lines = logContent.split(/\r?\n/).filter(l => l.trim() !== '');
  if (lines.length === 0) return { valid: true };

  let validLines = 0;
  for (const line of lines) {
    if (pattern.test(line) || line.startsWith('    at ') || line.startsWith('\tat ')) {
      validLines++;
    }
  }

  if ((validLines / lines.length) < 0.5) {
    return { valid: false, errorMessage: 'Log content does not match expected format' };
  }

  return { valid: true };
}

/**
 * Validates output size against template total size with fixed buffer
 */
export function isValidOutputSizeWithBuffer(htmlContent, templateTotalSize) {
  if (!htmlContent) return true;
  const outputSize = Buffer.byteLength(htmlContent, 'utf8');
  if (templateTotalSize > 0) {
    const maxAllowed = templateTotalSize + OUTPUT_SIZE_BUFFER;
    return outputSize <= maxAllowed;
  }
  return false; // if template size unknown, reject
}

/**
 * Get template total size for an appSite and appView from ConfigUtil
 */
export function getTemplateTotalSize(appSite, appView = '') {
  try {
    const scenarios = ConfigUtil.getScenarios();
    const match = scenarios.find(s =>
      s.appSite.toLowerCase() === appSite.toLowerCase() &&
      (s.appView || '').toLowerCase() === (appView || '').toLowerCase()
    );
    return match ? (match.totalSize || 0) : 0;
  } catch (e) {
    return 0;
  }
}
